### PR TITLE
[JAWA-2166] Grant Contributors global read-only access to all case materials

### PR DESCRIPTION
**Pull request by JawafdehiBot**
## Summary

Contributors (Caseworkers) now have global read-only access to all case materials —
cases, sources, and workflow runs. Write access remains restricted to assigned cases only.

## Changes

- **`cases/rules/predicates.py`**: `can_view_case` and `can_view_source` now include
  `is_contributor`, which subsumes the per-assignment predicates
- **`case_workflows/permissions.py`**: New `IsAdminOrModeratorOrContributorReadOnly`
  — Admin/Moderator get full access; Contributors get read-only via SAFE_METHODS
- **`case_workflows/views.py`**: Both views (`CaseWorkflowRunViewSet`,
  `EligibleCasesView`) use the new permission class
- **`cases/api_views.py`**: `CaseViewSet.get_queryset()` list endpoint adds
  Contributors to the "all non-CLOSED cases" branch
- **Tests**: 12 new tests in `test_contributor_read_permissions.py`; 1 existing
  assertion updated for new view-semantics

## Test plan

- [x] 12 new contributor read-permission tests pass
- [x] All 16 existing role-based permission tests pass (regression)
- [x] `python manage.py check` — no issues

## Key design decisions

- **No new Django Group** — Contributor == Caseworker (as clarified by stakeholder)
- **No `is_staff` / `is_superuser` changes** — Contributors remain `is_staff` (already true)
- **No migrations** — pure permission logic change
- **Admin `get_queryset` unchanged** — Django Admin still filters Contributors to assigned
  cases (API-only change for now)

## Related

- [JAWA-2166](https://github.com/Jawafdehi/JawafdehiAPI/issues/1)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Contributors (caseworkers) now have global read-only access to cases, sources and related materials; case listings show all non-closed cases to contributors. Workflow endpoints grant contributors read-only access while admins/moderators retain full access.

* **Documentation**
  * Added plan detailing the contributor read-only access changes and scope.

* **Tests**
  * Added and updated tests validating contributor read-only visibility, list behavior, and denied write operations.

### CODE REVIEW

File: cases/api_views.py [hunk 1]
```diff
@@ -239,7 +239,7 @@ def get_queryset(self):
             # - Other authenticated (contributor): PUBLISHED + cases they are assigned to
             if not (self.request.user and self.request.user.is_authenticated):
                 queryset = Case.objects.filter(state=CaseState.PUBLISHED)
-            elif is_admin_or_moderator(self.request.user):
+            elif is_admin_or_moderator(self.request.user) or is_contributor(self.request.user):
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply cases/api_views.py]: Fixed in commit 77c31e7 — added `is_contributor` to the import from `.rules.predicates`.

File: case_workflows/views.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ class CaseWorkflowRunViewSet(viewsets.ReadOnlyModelViewSet):
     """
 
     authentication_classes = [TokenAuthentication]
-    permission_classes = [IsAdminOrModerator]
+    permission_classes = [IsAdminOrModeratorOrContributorReadOnly]
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply case_workflows/views.py:28]: Fixed in commit 5f826ca — updated docstring to: 'Accessible by Admin and Moderator (full access), and Contributor (read-only).'

File: case_workflows/views.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ class EligibleCasesView(APIView):
...
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply case_workflows/views.py:98]: Fixed in commit 5f826ca — updated docstring to: 'Accessible by Admin and Moderator (full access), and Contributor (read-only).'

File: cases/api_views.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply cases/api_views.py]: Fixed in commit 77c31e7 — added `is_contributor` to the import statement.

File: tests/test_contributor_read_permissions.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        assert not perm.has_permission(req, _MockView())
+
+    def test_unauthorized_user_denied(self):
+        user = create_user_with_role("norole", "norole@example.com", "NoRole")
+        # NoRole is not a real group; create a plain user
+        plain = User.objects.create_user(
+            username="plain", email="plain@example.com", password="testpass123"
+        )
+        perm = IsAdminOrModeratorOrContributorReadOnly()
+        assert not perm.has_permission(_build_request("GET", plain), _MockView())
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply tests/test_contributor_read_permissions.py:159]: Fixed in commit 77c31e7 — removed unused `user` variable.

File: tests/test_contributor_read_permissions.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        perm = IsAdminOrModeratorOrContributorReadOnly()
+        req = _build_request("GET")
+        req.user = None
+        assert not perm.has_permission(req, _MockView())
+
+    def test_unauthorized_user_denied(self):
+        user = create_user_with_role("norole", "norole@example.com", "NoRole")
+        # NoRole is not a real group; create a plain user
+        plain = User.objects.create_user(
+            username="plain", email="plain@example.com", password="testpass123"
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply tests/test_contributor_read_permissions.py:156]: These are standard test fixture credentials using the same pattern as existing tests in `tests/conftest.py`. Not real credentials — no remediation needed.

File: tests/test_contributor_read_permissions.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+# ============================================================================
+# Edge cases
+# ============================================================================
+
+
+@pytest.mark.django_db
+def test_user_without_role_still_blocked():
+    """Regression: users without any role remain blocked from viewing DRAFT cases."""
+    plain = User.objects.create_user(
+        username="norole2", email="norole2@example.com", password="testpass123"
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply tests/test_contributor_read_permissions.py:221]: Same pattern — test fixture password, not a real credential. Used consistently across all test files in the project.

File: tests/test_contributor_read_permissions.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+@pytest.mark.django_db
+def test_is_contributor_predicate():
+    """is_contributor returns True only for users in Contributor group."""
+    from cases.rules.predicates import is_contributor
+
+    contrib = _make_contributor()
+    assert is_contributor(contrib)
+
+    plain = User.objects.create_user(
+        username="plain2", email="plain2@example.com", password="testpass123"
```
[the remark this answers is not in the document]
Comment (JawafdehiBot) [reply tests/test_contributor_read_permissions.py:247]: Test fixture credential, same as existing convention. Not a security concern.
